src/lib.rs

```diff
@@ -254,3 +265,334 @@ impl<'a, NP: ?Sized + NixPath>  NixPath for Option<&'a NP> {
+    /// when combined with `fork`).
+    ///
+    /// ```
+    /// use std::iter;
+    /// use std::ffi::CString;
+    /// use nix::{TerminatedVec, unistd};
+    /// use nix::sys::wait;
+    /// use nix::libc::_exit;
+    ///
+    /// # #[cfg(target_os = "android")]
+    /// # fn exe_path() -> CString {
+    /// # CString::new("/system/bin/sh").unwrap()
+    /// # }
+    /// # #[cfg(not(target_os = "android"))]
+    /// # fn exe_path() -> CString {
+    /// let exe = CString::new("/bin/sh").unwrap();
+    /// #    exe
+    /// # }
+    /// # let exe = exe_path();
+    /// let args = [
+    ///     exe.clone(),
+    ///     CString::new("-c").unwrap(),
+    ///     CString::new("echo hi").unwrap(),
+    /// ];
+    /// let args_p = TerminatedVec::terminate_cstr(&args);
+    /// let env = TerminatedVec::terminate(iter::empty());
+    ///
+    /// match unistd::fork().unwrap() {
+    ///     unistd::ForkResult::Child => {
+    ///         match unistd::execve(exe.as_c_str(), &args_p, &env) {
+    ///             Err(..) => {
+    ///                 // Panicking or returning will drop locals and
+    ///                 // deallocate memory, so let's avoid that here.
+    ///                 unsafe { _exit(1); }
+    ///             },
+    ///             Ok(..) => unreachable!(),
+    ///         }
+    ///     },
+    ///     unistd::ForkResult::Parent { child } => {
+    ///         let status = wait::waitpid(child, None).unwrap();
+    ///         assert_eq!(status, wait::WaitStatus::Exited(child, 0));
+    ///     },
+    /// }
+    /// ```
+    pub fn terminate_cstr<T: AsRef<CStr> + 'a, I: IntoIterator<Item=T>>(iter: I) -> Self {
+        fn cstr_char<'a, S: AsRef<CStr> + 'a>(s: S) -> &'a c_char {
+            unsafe {
+                &*s.as_ref().as_ptr()
+            }
+        }
+
+        Self::terminate(iter.into_iter().map(cstr_char))
+    }
+}
+
+impl<T> Deref for TerminatedVec<T> {
+    type Target = TerminatedSlice<T>;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe {
+            TerminatedSlice::from_slice_unchecked(&self.inner)
+        }
+    }
+}
+
+impl<T> DerefMut for TerminatedVec<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe {
+            TerminatedSlice::from_slice_mut_unchecked(&mut self.inner)
+        }
+    }
+}
+
+impl<T> AsRef<TerminatedSlice<T>> for TerminatedVec<T> {
+    fn as_ref(&self) -> &TerminatedSlice<T> {
+        self
+    }
+}
+
+impl<'a, T: 'a> IntoRef<'a, TerminatedSlice<T>> for TerminatedVec<T> {
+    type Target = TerminatedVec<T>;
+
+    fn into_ref(self) -> Self::Target {
+        self
+    }
+}
+
+impl<'a, T> IntoRef<'a, TerminatedSlice<T>> for &'a TerminatedVec<T> {
+    type Target = &'a TerminatedSlice<T>;
+
+    fn into_ref(self) -> Self::Target {
+        self
+    }
+}
+
+/// Coercion of `CStr` iterators into an argument that can be passed to `exec`.
+impl<'a, T: AsRef<CStr> + 'a, I: IntoIterator<Item=T>> IntoRef<'a, TerminatedSlice<&'a c_char>> for I {
```

Comment:
Is this something that makes sense to have globally-defined? Or can this be integrated into the function definition instead?

Reply:
This impl is pretty central to allowing exec to accept either:

- `&TerminatedSlice` (usually actually a `TerminatedVec` created by `terminate()`)
- `&[CStr]` and similar (specifically any `IntoIterator` where `Item: AsRef<CStr>`)

... particularly the latter, which enables the old behaviour of allocate-at-call-site. One slightly less global approach might be to move the `IntoRef` trait and its impls into the `unistd` module and possibly give it a name that isn't as generic sounding.

Reply:
Another option would be to simply remove the old style and require the use of `TerminatedVec::terminate_cstr` any time exec is called, whether the Vec is being preallocated or not.

Reply:
I'm a big fan of reducing options and improving consistency. This is why I've been asking for code examples, as I'm not familiar with how people use this API regularly. I want to make sure we know how people use it regularly and our API is easy for the 90% case and possible for the last 10% case. Or something along those lines.

All the examples I see involve a hard-coded array. And in those situations, I think a `term_vec!()` macro makes a lot of sense; this would look like a regular `vec!` call and avoid all of the wrapping in `Option`. But I don't know how people build up these arrays programmatically or if that's ever done in practice. Is that a common operation? I'd think that is handled nicely by the API provided here, but I don't know exactly.

Reply:
Well, examples are going to use a hard-coded array because they're just examples. Real world usage varies of course, and can be hard-coded in simple cases, or generated from a config file or argv [like in the use case that I originally wrote these changes for](https://github.com/arcnmx/encage-runtime/blob/master/src/main.rs#L232).

Most uses for exec I can imagine would use data coming from a dynamic source rather than being hard-coded. The convenient no-preallocation variant mostly comes into play when using exec in a program that doesn't fork and just intends to replace the whole process - I'm not sure how common that is, I guess if you're writing a tool like [env(1)](https://linux.die.net/man/1/env) or something.